### ISSUE TITLE
Live output is incorrect with filters and triggers

### ISSUE THREAD

**Issue by clementguidi** [namhyung/uftrace#1589]
Hi,

I noticed an incorrect output when using the `live` command with filters and triggers enabled. The output from `live` differs from the output from `record` followed by `replay`.

Example:
```
$ cat abc.c
void c(void) {
	/* do nothing */
}

void b(void) {
	c();
}

void a(void) {
	b();
}

int main(void) {
	a();
	return 0;
}
$ gcc -pg -o abc abc.c

$ uftrace -T a@depth=1 -F. abc
# DURATION     TID     FUNCTION
   1.397 us [490528] | __monstartup();
   1.467 us [490528] | __cxa_atexit();
            [490528] | main() {
   6.565 us [490528] |   a();
   8.520 us [490528] | } /* main */

$ uftrace record -T a@depth=1 -F. abc
$ uftrace replay
# DURATION     TID     FUNCTION
   1.397 us [490534] | __monstartup();
   1.466 us [490534] | __cxa_atexit();
            [490534] | main() {
            [490534] |   a() {
            [490534] |     b() {
   1.397 us [490534] |       c();
   3.771 us [490534] |     } /* b */
   6.145 us [490534] |   } /* a */
   8.520 us [490534] | } /* main */
```

The `-F` option doesn't apply with `live`, as filtered functions should appear despite the trigger.

Maybe the options are not handled correctly? I didn't take time to investigate.

**Comment by namhyung** [comment]
Hmm.. I believe the record + replay behavior is correct, and it seems the live command processed filter incorrectly.  I think it skips some filters during live-replay.. will check.

**Comment by honggyukim** [comment]
Hmm... I think the output of `live` command looks correct and `record` + `replay` looks weird.

@namhyung Do you really think the output of `live` command is incorrect?

**Comment by namhyung** [comment]
`-F .` enables every function to be traced.  Without that it would work as live command.  Actually it has `reset_live_opts()` to reset filter options for report since it's already handled in record.  But it didn't reset trigger options because it might have replay-time actions too.

**Comment by namhyung** [comment]
Something like [below patch](https://github.com/namhyung/uftrace/files/9743419/uftrace-live-reset-trigger-option.patch.txt) would be needed:

```diff
diff --git a/cmds/live.c b/cmds/live.c
index f3620d657b..3d8e659df5 100644
--- a/cmds/live.c
+++ b/cmds/live.c
@@ -35,6 +35,65 @@ static void reset_live_opts(struct uftrace_opts *opts)
        free(opts->caller);
        opts->caller = NULL;
 
+       /*
+        * Likewise, most trigger options are ignored, but color settings
+        * are effective.
+        */
+       if (opts->trigger) {
+               char *new_triggers = NULL;
+               struct strv trs = STRV_INIT;
+               char *s;
+               int i;
+
+               if (strstr(opts->trigger, "color=") == NULL) {
+                       free(opts->trigger);
+                       opts->trigger = NULL;
+                       goto others;
+               }
+
+               /*
+                * Split trigger options for each function.
+                * Example trigger option string like this:
+                *
+                *  a@depth=1;b@trace,read=pmu-cycle;c@color=red,time=1us
+                */
+               strv_split(&trs, opts->trigger, ";");
+
+               strv_for_each(&trs, s, i) {
+                       struct strv sv = STRV_INIT;
+                       char *name, *tmp, *o;
+                       int k;
+
+                       if (strstr(s, "color=") == NULL)
+                               continue;
+
+                       name = s;
+                       tmp = strchr(s, '@');
+                       if (tmp == NULL) {
+                               pr_dbg("invalid trigger option: %s\n", s);
+                               continue;
+                       }
+                       *tmp = '\0';
+
+                       /* split trigger option into actions */
+                       strv_split(&sv, tmp + 1, ",");
+
+                       strv_for_each(&sv, o, k) {
+                               if (strncmp(o, "color=", 6))
+                                       continue;
+
+                               tmp = strjoin(name, o, "@");
+                               new_triggers = strjoin(new_triggers, tmp, ";");
+                               break;
+                       }
+               }
+               strv_free(&trs);
+
+               free(opts->trigger);
+               opts->trigger = new_triggers;
+       }
+
+others:
        opts->depth = MCOUNT_DEFAULT_DEPTH;
        opts->disabled = false;
        opts->no_event = false;
```

**Comment by namhyung** [comment]
Merged 52c7bd3cc3.

```
$ uftrace -T a@depth=1 -F.  t-abc
# DURATION     TID     FUNCTION
   1.838 us [897263] | __monstartup();
   0.959 us [897263] | __cxa_atexit();
            [897263] | main() {
            [897263] |   a() {
            [897263] |     b() {
            [897263] |       c() {
   2.060 us [897263] |         getpid();
   3.376 us [897263] |       } /* c */
   3.986 us [897263] |     } /* b */
   4.529 us [897263] |   } /* a */
   5.307 us [897263] | } /* main */
```
